src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp

```diff
@@ -185,13 +258,53 @@ void Gather<dim>::operator()(container::CartBuffer<Type, memDim, T_Alloc, T_Copy
     if(useTmpSrc)
         tmpSrc = source; /* Mem copy */
 
+    // Get number of elements for each source buffer
+    std::vector<Size_t<memDim> > srcBufferSizes(numRanks);
+    Size_t<memDim> srcBufferSize = source.size();
     MPI_CHECK(MPI_Gather(
-               useTmpSrc ? (void*)tmpSrc.getDataPointer(): (void*)source.getDataPointer(), source.size().productOfComponents() * sizeof(Type), MPI_CHAR,
-               root() ? (void*)tmpDest.data() : NULL, source.size().productOfComponents() * sizeof(Type), MPI_CHAR,
+        (void*)&srcBufferSize,
+        sizeof(Size_t<memDim>),
+        MPI_CHAR,
+        (void*)srcBufferSizes.data(),
+        sizeof(Size_t<memDim>),
+        MPI_CHAR,
+        0, this->comm));
+
+    // 1D offsets in destination buffer
+    std::vector<size_t> srcBufferOffsets1D(numRanks);
+    std::vector<size_t> srcBufferSizes1D(numRanks);
+    std::vector<int> srcBufferOffsets1D_char(numRanks);
```

Comment:
Can you then add assertions that the sizes are not above the integer range? Would be line 290f. where the casts are.